planner/core/rule_join_elimination.go

```diff
@@ -0,0 +1,165 @@
+	cols    [][]*expression.Column
+	schemas []*expression.Schema
+}
+
+// tryToEliminateOuterJoin will eliminate outer join plan base on the following rules
+// 1. outer join elimination: For example left outer join, if the parent only use the
+//    columns from left table and the join key of right table(the inner table) is a unique
+//    key of the right table. the left outer join can be eliminated.
+// 2. outer join elimination with distinct: For example left outer join. If the parent
+//    only use the columns from left table with 'distinct' label. The left outer join can
+//    be eliminated.
+func (o *outerJoinEliminator) tryToEliminateOuterJoin(p *LogicalJoin) LogicalPlan {
+	switch p.JoinType {
+	case LeftOuterJoin:
+		return o.doEliminate(p, 1)
+	case RightOuterJoin:
+		return o.doEliminate(p, 0)
+	default:
+		return nil
+	}
+}
+
+func (o *outerJoinEliminator) doEliminate(p *LogicalJoin, innerChildIdx int) LogicalPlan {
+	// outer join elimination with distinct
+	if len(o.cols) > 0 {
+		cols := o.cols[len(o.cols)-1]
+		allColsInSchema := true
+		for _, col := range cols {
+			columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+			if c, _ := p.children[1^innerChildIdx].Schema().FindColumn(columnName); c == nil {
+				allColsInSchema = false
+				break
+			}
+		}
+		if allColsInSchema == true {
+			return p.children[1^innerChildIdx]
+		}
+	}
+
+	// outer join elimination without distinct
+	// first, check whether the parent's schema columns are all in left or right
+	if len(o.schemas) == 0 {
+		return nil
+	}
+	for _, col := range o.schemas[len(o.schemas)-1].Columns {
+		columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+		if c, _ := p.children[1^innerChildIdx].Schema().FindColumn(columnName); c == nil {
+			return nil
+		}
+	}
+	// second, check whether the other side join keys are unique keys
+	p.children[innerChildIdx].buildKeyInfo()
+	var joinKeys []*expression.Column
+	for _, eqCond := range p.EqualConditions {
+		joinKeys = append(joinKeys, eqCond.GetArgs()[innerChildIdx].(*expression.Column))
+	}
+	tmpSchema := expression.NewSchema(joinKeys...)
+	for _, keyInfo := range p.children[innerChildIdx].Schema().Keys {
+		joinKeysContainKeyInfo := true
+		for _, col := range keyInfo {
+			columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+			if c, _ := tmpSchema.FindColumn(columnName); c == nil {
+				joinKeysContainKeyInfo = false
+				break
+			}
+		}
+		if joinKeysContainKeyInfo {
+			return p.children[1^innerChildIdx]
+		}
+	}
+	// Third, if p.children[innerChildIdx] is datasource, we must check specially index.
+	// Because buildKeyInfo() don't save the index without notnull flag.
+	// But in outer join, null==null don't return true. The null index do not affect the res.
+	if ds, ok := p.children[innerChildIdx].(*DataSource); ok {
+		for _, path := range ds.possibleAccessPaths {
+			if path.isTablePath {
```

Comment:
there are some unique table paths, for example int primary key, the outer join in the following example can also be eliminated:
```sql
TiDB(localhost:4000) > create table t1(a bigint, b bigint);
Query OK, 0 rows affected (0.01 sec)

TiDB(localhost:4000) > create table t2(a bigint primary key, b bigint);
Query OK, 0 rows affected (0.01 sec)

TiDB(localhost:4000) > desc select * from t1 left outer join t2 on t1.a=t2.a and t1.b=t2.b;
+------------------------+----------+------+--------------------------------------------------------------------------------------------------+
| id                     | count    | task | operator info                                                                                    |
+------------------------+----------+------+--------------------------------------------------------------------------------------------------+
| HashLeftJoin_9         | 12500.00 | root | left outer join, inner:TableReader_13, equal:[eq(test.t1.a, test.t2.a) eq(test.t1.b, test.t2.b)] |
| ├─TableReader_11       | 10000.00 | root | data:TableScan_10                                                                                |
| │ └─TableScan_10       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                                      |
| └─TableReader_13       | 10000.00 | root | data:TableScan_12                                                                                |
|   └─TableScan_12       | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                                      |
+------------------------+----------+------+--------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

Reply:
Probably `select *` can't be eliminated?

Reply:
Oh, you're right.

Reply:
Will not eliminate. Because t1.a is not a unique key.

Reply:
```
tidb(localhost:4000) > explain select t1.a, t1.b from t1 left outer join t2 on t1.a = t2.a;
+-------------------+----------+------+-------------------------------------------------------------+
| id                | count    | task | operator info                                               |
+-------------------+----------+------+-------------------------------------------------------------+
| TableReader_7     | 10000.00 | root | data:TableScan_6                                            |
| └─TableScan_6     | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo |
+-------------------+----------+------+-------------------------------------------------------------+
2 rows in set (0.00 sec)
```
It can be eliminated.